### PR TITLE
changefeedccl: use new parquet library

### PR DESCRIPTION
**Pull request by jayshrivastava**
This change updates changefeeds to use the new parquet library
added in `pkg/util/parquet` when using `format=parquet`.
    
Informs: https://github.com/cockroachdb/cockroach/issues/99028
Epic: https://cockroachlabs.atlassian.net/browse/CRDB-15071
Release note: None

### CODE REVIEW

**Review by jayshrivastava — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103293)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[`pkg/ccl/changefeedccl/changefeedbase/options.go` line 1061 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/103293#-NVW0QtC-zSdQ3Ki23cu:-NVW0QtC-zSdQ3Ki23cv:b-4qziez) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ceb218e2fbc92413344d041682b197228f183ea/pkg/ccl/changefeedccl/changefeedbase/options.go#L1061)):*
> ```Go
> 	// Right now parquet does not support any of these options
> 	if s.m[OptFormat] == string(OptFormatParquet) {
> 		if isPredicateChangefeed {
> ```

Enabling parquet + predicate changefeeds + diff was added in https://github.com/cockroachdb/cockroach/pull/94653, but there's not much of an explanation for why. If you look at the `parquetPossible` randomized logic in `pkg/ccl/changefeedccl/testfeed_test.go`, it looks like we disabled testing the parquet format with `diff`. I've decided to leave `diff` unsupported until we add it explicitly with proper testing in https://github.com/cockroachdb/cockroach/issues/103129.

File: pkg/ccl/changefeedccl/parquet.go [hunk 1]
```diff
@@ -37,27 +38,46 @@ func newParquetWriterFromRow(
...
+) (*parquetWriter, error) {
+	schemaDef, numCols, err := newParquetSchemaDefintion(row)
 	if err != nil {
 		return nil, err
 	}
 
 	writerConstructor := parquet.NewWriter
+
 	if includeParquetTestMetadata {
+		if opts, err = addCDCTestMetadata(row, opts); err != nil {
```
Comment (miretskiy) [inline pkg/ccl/changefeedccl/parquet.go:67]: I wonder if we can get rid of includeParquetTestMetadata altogether? LIke at the call site (testfeed?) if we decide to use parquet, then.. just explicitly specify additional metadata?

File: pkg/util/parquet/schema.go [hunk 2]
```diff
@@ -341,6 +344,10 @@ func makeColumn(colName string, typ *types.T, repetitions parquet.Repetition) (c
 		// and [] when encoding.
 		// There is more info about encoding arrays here:
 		// https://arrow.apache.org/blog/2022/10/08/arrow-parquet-encoding-part-2/
+		if typ.ArrayContents().Family() == types.ArrayFamily {
+			return result, pgerror.Newf(pgcode.FeatureNotSupported,
+				"parquet writer does not support nested arrays")
+		}
```
Comment (miretskiy) [inline pkg/util/parquet/schema.go:350]: why not?

**Review by miretskiy — COMMENTED** [review]
I think we need to address the fact that tuples are not supported; this breaks cdc queries, and cdc_prev tuple.
cdc_prev also turns on "diff" option.

**Review by jayshrivastava — COMMENTED** [review]
I think this is okay to merge. Supporting tuples (needed for cdc_prev) is on the list of items to complete before making the new library public https://github.com/cockroachdb/cockroach/issues/99028. I'm working on those changes concurrently to this PR and plan to merge them next. There are some issues with the old library, so my changes are strictly better than what is in the main branch IMO:

The old library "supports" cdc_prev, tuples, and diff, but they don't work. You can create a changefeed ex. `create changefeed into 'nodelocal://0/baz' with format='parquet', diff as select *, cdc_prev from foo;`. The `cdc_prev` tuple causes an error, because tuples are not supported. If you pass `diff`, parquet does not emit the diff. The reason for this surprising behavior is that we disable parquet in tests that use `diff`, but enable parquet in production with `diff` when using cdc queries. My change explicitly disallows these cases in production and testing. I will incrementally enable them as I add more features.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103293#-:-NVZjsi3BWTDmFAYzdU7:br8ekff)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB and @miretskiy)

___
*[`pkg/ccl/changefeedccl/parquet.go` line 67 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/103293#gh-1194326290:-NVZmW-Q47ApztaQGt3b:b4njhgt) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddb476239ccddd0e1006866b76955d1da54fc08b/pkg/ccl/changefeedccl/parquet.go#L67)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I wonder if we can get rid of includeParquetTestMetadata altogether? LIke at the call site (testfeed?) if we decide to use parquet, then.. just explicitly specify additional metadata?
</blockquote></details>

Can you please clarify this?  The metadata contains information from the rows (the keys and non keys) which are used in testing ex. ```assertPayloads([]string{`[key] -> {"a": 1}`})```. I think this function is the best place to get that information easily. 
___
*[`pkg/util/parquet/schema.go` line 350 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/103293#gh-1194328711:-NVZndFk9Ox6YWD2O4cy:btrzjy7) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddb476239ccddd0e1006866b76955d1da54fc08b/pkg/util/parquet/schema.go#L350)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

why not?
</blockquote></details>

A few reasons.

If I understand correctly, we don't support multidimensional arrays, even in projections, so we should never see them. @HonoreDB also tried a few cases with me, but we couldn't make a nested array.
```
select array[[1,2,3],[4,5,6]] as col;                                                                                                                                                                       
ERROR: unimplemented: arrays cannot have arrays as element type
SQLSTATE: 0A000
HINT: You have attempted to use a feature that is not yet implemented.
See: https://go.crdb.dev/issue-v/32552/v23.1
```

The code I wrote definitely does not support nested arrays, it assumes arrays are arrays of scalar, so I just made the assertion explicit.

I'm fairly confident arbitrary nesting is not supported in parquet because I could not find anything about it in the docs/spec or by searching online. At best, there are some hacks (https://medium.com/hadoop-noob/recursive-avro-schema-for-parquet-a87c84b6aca1), but I don't feel confident relying on them. Chatgpt does not recommend it either :)

**Review by miretskiy — APPROVED** [review]
Okay; just for my info: what kind of error?
If you don't have explicit issue, could you file one to get cdc_prev/diff/parquet working in a sane way?

> Reviewed 1 of 13 files at r1, 1 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103293#-:-NV_164ND9ihC27LeDDi:b-2farni)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB and @jayshrivastava)

___
*[`pkg/ccl/changefeedccl/parquet.go` line 114 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103293#-NV_3Tbp0a7oM1QdoRZO:-NV_3Tbp0a7oM1QdoRZP:b-th0b1y) ([raw file](https://github.com/cockroachdb/cockroach/blob/82ba41f1ba3721889f124b7c29092c204c72c91c/pkg/ccl/changefeedccl/parquet.go#L114)):*
> ```Go
> // addCDCTestMetadata appends options to the provided options to configure the
> // parquet writer to write metadata required by cdc test feed factories.
> func addCDCTestMetadata(row cdcevent.Row, opts []parquet.Option) ([]parquet.Option, error) {
> ```

nit: perhaps addParquetTestMetadata?
___
*[`pkg/ccl/changefeedccl/parquet.go` line 67 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/103293#gh-1194326290:-NV_1GcK83FlRMEIiDAs:b-qx8nf3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddb476239ccddd0e1006866b76955d1da54fc08b/pkg/ccl/changefeedccl/parquet.go#L67)):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

Can you please clarify this?  The metadata contains information from the rows (the keys and non keys) which are used in testing ex. ```assertPayloads([]string{`[key] -> {"a": 1}`})```. I think this function is the best place to get that information easily. 
</blockquote></details>

I was just curious if we can explicitly pass an argument indicating whether or not we should include test metadata.
Perhaps rely on some knobs -- like if changefeed knob is non-nil, then this is test code, and so, add this metdata.
I understand why we need this metadata -- just not sure we need this global boolean flag to achieve this.
Feel free to ignore if too much hassle.
___
*[`pkg/ccl/changefeedccl/changefeedbase/options.go` line 1061 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/103293#-NVW0QtC-zSdQ3Ki23cu:-NV_1sMe1m57NnnwzjGT:b-tsbkal) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ceb218e2fbc92413344d041682b197228f183ea/pkg/ccl/changefeedccl/changefeedbase/options.go#L1061)):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

Enabling parquet + predicate changefeeds + diff was added in https://github.com/cockroachdb/cockroach/pull/94653, but there's not much of an explanation for why. If you look at the `parquetPossible` randomized logic in `pkg/ccl/changefeedccl/testfeed_test.go`, it looks like we disabled testing the parquet format with `diff`. I've decided to leave `diff` unsupported until we add it explicitly with proper testing in https://github.com/cockroachdb/cockroach/issues/103129. 
</blockquote></details>

As I recall it, it was made purely to make tests work -- we randomly pick parquet sink sometimes.
___
*[`pkg/util/parquet/schema.go` line 350 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/103293#gh-1194328711:-NV_1yIQE6yptxZP4GLg:bw4egza) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddb476239ccddd0e1006866b76955d1da54fc08b/pkg/util/parquet/schema.go#L350)):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

A few reasons.

If I understand correctly, we don't support multidimensional arrays, even in projections, so we should never see them. @HonoreDB also tried a few cases with me, but we couldn't make a nested array.
```
select array[[1,2,3],[4,5,6]] as col;                                                                                                                                                                       
ERROR: unimplemented: arrays cannot have arrays as element type
SQLSTATE: 0A000
HINT: You have attempted to use a feature that is not yet implemented.
See: https://go.crdb.dev/issue-v/32552/v23.1
```

The code I wrote definitely does not support nested arrays, it assumes arrays are arrays of scalar, so I just made the assertion explicit.

I'm fairly confident arbitrary nesting is not supported in parquet because I could not find anything about it in the docs/spec or by searching online. At best, there are some hacks (https://medium.com/hadoop-noob/recursive-avro-schema-for-parquet-a87c84b6aca1), but I don't feel confident relying on them. Chatgpt does not recommend it either :) 
</blockquote></details>

Ack. Thanks for the explanation.  PostgreSQL supports multi dimensional arrays, and I think it's a matter of time before we do too.

I wonder if we should have a fall back mechanism i this library: if unsure -- convert to JSON.  Every type must be convertable to json.

At any array, could you add a bit of a commentary here, and link to
https://go.crdb.dev/issue-v/32552/v23.1
___
*[`pkg/ccl/changefeedccl/testfeed_test.go` line 1347 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103293#-NV_32j5EfOczUcGCh8p:-NV_32j5EfOczUcGCh8q:b-l9pyfa) ([raw file](https://github.com/cockroachdb/cockroach/blob/82ba41f1ba3721889f124b7c29092c204c72c91c/pkg/ccl/changefeedccl/testfeed_test.go#L1347)):*
> ```Go
> 		}
> 
> 		keyBuf := bytes.Buffer{}
> ```

`var keyBuf bytes.buffer`?

**Review by jayshrivastava — COMMENTED** [review]
The error is `transient error: parquet export does not support the TupleFamily type yet`. At least it's a sane error. With this change, you won't even be able to create the changefeed. I'm working on adding tuples now - https://github.com/cockroachdb/cockroach/pull/103589. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103293#-:-NVjC6VG6WLuWNgsVXtV:bio77yg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB and @miretskiy)

___
*[`pkg/ccl/changefeedccl/parquet.go` line 114 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103293#-NV_3Tbp0a7oM1QdoRZO:-NVjILoZB9VPqIuogLK5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/82ba41f1ba3721889f124b7c29092c204c72c91c/pkg/ccl/changefeedccl/parquet.go#L114)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

nit: perhaps addParquetTestMetadata?
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/parquet.go` line 67 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/103293#gh-1194326290:-NVjIJk-7yhKvCv7LboB:b-bhlyb5) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddb476239ccddd0e1006866b76955d1da54fc08b/pkg/ccl/changefeedccl/parquet.go#L67)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I was just curious if we can explicitly pass an argument indicating whether or not we should include test metadata.
Perhaps rely on some knobs -- like if changefeed knob is non-nil, then this is test code, and so, add this metdata.
I understand why we need this metadata -- just not sure we need this global boolean flag to achieve this.
Feel free to ignore if too much hassle.
</blockquote></details>

Done. I made it a testing knob.
___
*[`pkg/util/parquet/schema.go` line 350 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/103293#gh-1194328711:-NVjIN4B7HfM7pEufy1w:b9g0lyz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddb476239ccddd0e1006866b76955d1da54fc08b/pkg/util/parquet/schema.go#L350)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Ack. Thanks for the explanation.  PostgreSQL supports multi dimensional arrays, and I think it's a matter of time before we do too.

I wonder if we should have a fall back mechanism i this library: if unsure -- convert to JSON.  Every type must be convertable to json.

At any array, could you add a bit of a commentary here, and link to
https://go.crdb.dev/issue-v/32552/v23.1
</blockquote></details>

Ack. Will implement this as discussed IRL. Working on adding tuple support as well.
___
*[`pkg/ccl/changefeedccl/testfeed_test.go` line 1347 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103293#-NV_32j5EfOczUcGCh8p:-NVjIJ5x93tQyfT9G96j:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/82ba41f1ba3721889f124b7c29092c204c72c91c/pkg/ccl/changefeedccl/testfeed_test.go#L1347)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

`var keyBuf bytes.buffer`?
</blockquote></details>

Done.

**Comment by jayshrivastava** [comment]
bors r=miretskiy

**Pull request merged**